File: bst/cpp/lb-01-bst-insert-search.cpp
#include <iostream>
using namespace std;

class Node {
   public:
    int data;
    Node *left, *right;
    Node(int val) : data(val), left(NULL), right(NULL) {}
};
class BST {
   public:
    Node *root;
    BST() : root(NULL) {}
    Node *insert_helper(Node *root, int val) {
        if (!root) {
            return new Node(val);
        }
        if (root->data == val)
            return root;
        else if (root->data > val) {
            root->left = insert_helper(root->left, val);
        } else {
            root->right = insert_helper(root->right, val);
        }
        return root;
    }
    Node *search_helper(Node *root, int val) {
        if (!root) {
            return NULL;
        }
        if (root->data == val)
            return root;
        else if (root->data > val) {
            return search_helper(root->left, val);
        }
        return search_helper(root->right, val);
    }
    Node *insert(int val) {
        // assign to root as root may not exits.
        root = insert_helper(root, val);
        return root;
    }
    Node *search(int val) {
        return search_helper(root, val);
    }
};

int main() {
    int arr[] = {4, 2, 7, 1, 3};
    int n = sizeof(arr) / sizeof(int);
    BST b = BST();
    for (auto x : arr) {
        b.insert(x);
    }
    for (auto x : arr) {
        cout << x << "=" << b.search(x)->data << ", ";
    }
    cout << b.search(11);
    return 0;
}

File: bst/lb-01-bst-insert-search.py

class Node:
    def __init__(self, val):
        self.data = val
        self.left = None
        self.right = None


class BST:
    def __init__(self):
        self.root = None

    def insert(self, val):

        def helper(root, val):
            if not root:
                return Node(val)

            if root.data == val:
                return root

            elif root.data < val:
                root.right = helper(root.right, val)

            elif root.data > val:
                root.left = helper(root.left, val)

            return root

        self.root = helper(self.root, val)

    def search(self, val):
        def helper(root, val):
            if not root:
                return -1
            if root.data == val:
                return root
            elif root.data < val:
                return helper(root.right, val)
            return helper(root.left, val)

        return helper(self.root,  val)


nodes = [4, 2, 7, 1, 3]
search = 8
bst = BST()
for ele in nodes:
    bst.insert(ele)
found = bst.search(search)
print(found.data == search if found != -1 else False)
